state/watcher.go

```diff
@@ -3221,12 +3221,133 @@ func ensureSuffixFn(marker string) func(string) string {
 // watchEnqueuedActionsFilteredBy starts and returns a StringsWatcher
 // that notifies on new Actions being enqueued on the ActionRecevers
 // being watched.
-func (st *State) watchEnqueuedActionsFilteredBy(receivers ...ActionReceiver) StringsWatcher {
-	return newCollectionWatcher(st, colWCfg{
-		col:    actionNotificationsC,
-		filter: makeIdFilter(st, actionMarker, receivers...),
-		idconv: actionNotificationIdToActionId,
+func (st *State) watchEnqueuedActionsFilteredBy(watchChanges bool, receivers ...ActionReceiver) StringsWatcher {
+	if watchChanges {
+		return newCollectionWatcher(st, colWCfg{
+			col:    actionNotificationsC,
+			filter: makeIdFilter(st, actionMarker, receivers...),
+			idconv: actionNotificationIdToActionId,
+		})
+	}
+	return newLegacyActionNotificationWatcher(st, receivers...)
+}
+
+// collectionWatcher is a StringsWatcher that watches for changes on the
+// specified collection that match a filter on the id.
+type legacyActionNotificationWatcher struct {
+	commonWatcher
+	source chan watcher.Change
+	sink   chan []string
+	filter func(interface{}) bool
+}
+
+// newLegacyActionNotificationWatcher starts and returns a new StringsWatcher configured
+// with the given collection and filter function
+func newLegacyActionNotificationWatcher(backend modelBackend, receivers ...ActionReceiver) StringsWatcher {
```

Comment:
I wonder if we can't unify the legacy and new watchers - the legacy watcher would only trigger if the status is pending; the new watcher would trigger if the status is pending or aborting. We'd have a single watcher with a filter used to expose the relevant changes.